bundles/org.openhab.core/src/main/java/org/openhab/core/cache/ByteArrayFileCache.java

```diff
@@ -0,0 +1,304 @@
+import java.math.BigInteger;
+import java.nio.file.Files;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.OpenHAB;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a simple file based cache implementation.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public class ByteArrayFileCache {
+
+    private final Logger logger = LoggerFactory.getLogger(ByteArrayFileCache.class);
+
+    static final String CACHE_FOLDER_NAME = "cache";
+    private static final char EXTENSION_SEPARATOR = '.';
+
+    private final File cacheFolder;
+
+    static final long ONE_DAY_IN_MILLIS = TimeUnit.DAYS.toMillis(1);
+    private long expiry;
+
+    private static final Map<String, File> FILES_IN_CACHE = new ConcurrentHashMap<>();
+
+    /**
+     * Creates a new {@link ByteArrayFileCache} instance for a service. Creates a <code>cache</code> folder under
+     * <code>$OPENHAB_USERDATA/cache/$servicePID</code>.
+     *
+     * @param servicePID PID of the service
+     */
+    public ByteArrayFileCache(String servicePID) {
+        // TODO track and limit folder size
+        // TODO support user specific folder
+        cacheFolder = new File(new File(OpenHAB.getUserDataFolder(), CACHE_FOLDER_NAME), servicePID);
+        if (!cacheFolder.exists()) {
+            logger.debug("Creating cache folder '{}'", cacheFolder.getAbsolutePath());
+            cacheFolder.mkdirs();
+        }
+        logger.debug("Using cache folder '{}'", cacheFolder.getAbsolutePath());
+    }
+
+    /**
+     * Creates a new {@link ByteArrayFileCache} instance for a service. Creates a <code>cache</code> folder under
+     * <code>$OPENHAB_USERDATA/cache/$servicePID/</code>.
+     *
+     * @param servicePID PID of the service
+     * @param long the days for how long the files stay in the cache valid. Must be positive. 0 to
+     *            disable this functionality.
+     */
+    public ByteArrayFileCache(String servicePID, long expiry) {
+        this(servicePID, Duration.ofDays(expiry));
+    }
+
+    /**
+     * Creates a new {@link ByteArrayFileCache} instance for a service. Creates a <code>cache</code> folder under
+     * <code>$OPENHAB_USERDATA/cache/$servicePID/</code>.
+     *
+     * @param servicePID PID of the service
+     * @param Duration the days for how long the files stay in the cache valid. Must be positive. 0 to
+     *            disable this functionality.
+     */
+    public ByteArrayFileCache(String servicePID, Duration expiry) {
+        this(servicePID);
+        if (expiry.isNegative()) {
+            throw new IllegalArgumentException("Cache expiration time must be greater than or equal to 0");
+        }
+        this.expiry = expiry.toDays();
```

Comment:
If we do a `expiry.toMillis()`, then we do not need to do the multiplication each time we check the expiry of a file.

And `duration` could be anything, not just days.